### ISSUE TITLE
Why cosine loss instead of just dot product?

### ISSUE THREAD

**Issue by ekgren** [openai/CLIP#68]
Trying to find the motivation in the article but I'm left without a definitive answer.

 What is the motivation of using L2 normed dot product instead of simply the dot product in the loss?

**Comment by jongwook** [comment]
The main purpose is to stabilize training, by limiting the dynamic range of the logits. With the cosine similarities being within [-1, 1] and an upper-bounded temperature parameter, it's a simple way to prevent the model predictions from exploding. [Radovanovi et al.](https://www.jmlr.org/papers/v11/radovanovic10a.html) found that the "hubness phenomenon" is less severe with cosine similarity, although synthetic data was used.
